### PR TITLE
add tests for ListContainerStats

### PR DESCRIPTION
**Pull request by minbaev**
Signed-off-by: Alexander Minbaev <alexander.minbaev@ibm.com>

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
PR adds tests for `ListContainerStats` RPC
Adds tests covering the RPC behavior after this change:
 https://github.com/containerd/containerd/pull/6373

#### Special notes for your reviewer:
None

#### Does this PR introduce a user-facing change?
None

### CODE REVIEW

File: pkg/validate/container.go
Comment (mikebrow) [inline pkg/validate/container.go:665]: s/listContainersStats/listContainerStats/

File: pkg/validate/container.go
Comment (mikebrow) [inline pkg/validate/container.go:664]: comment func name should match

**Review by mikebrow — APPROVED** [review]
/LGTM with nits addressed after we have the fix in containerd merged into main, and cherry picked to containerd 1.5.9 should only have to "pause" for this test case for a short period. If it takes to long for the 1.5.9 release to cycle up over this holiday feel free to add a skip.

**Comment by feiskyer** [comment]
could rebase again (it seems rebase is required to get new changes for github actions)?

**Comment by minbaev** [comment]
> could rebase again (it seems rebase is required to get new changes for github actions)?

sure, done

**Comment by minbaev** [comment]
@feiskyer @mikebrow 
can you please review and see if it can be merged?
Thank you!

**Comment by saschagrunert** [comment]
@minbaev the tests seems to fail, can we fix them?

**Comment by minbaev** [comment]
> @minbaev the tests seems to fail, can we fix them?

hmm, I don't think the tests are wrong, they pass when i run them locally against a local containerd instance with the change.

we cut a 1.5.9 containerd version which should include the change
https://github.com/containerd/containerd/pull/6373

can it be possible that the tests fail due some CI issue?


Here is the output when I run the tests locally
```
root@ubuntu-bm:~/go/src/github.com/kubernetes-sigs/cri-tools# build/bin/critest -ginkgo.focus stats
critest version: 1.22.0-79-g8aec4be0
Running Suite: CRI validation
=============================
Random Seed: 1642586400 - Will randomize all specs
Will run 4 of 89 specs

Jan 19 04:00:00.504: INFO: Testing context container image list file:
Jan 19 04:00:00.504: INFO: Testing context container image list: {DefaultTestContainerImage:busybox:1.28 WebServerTestImage:}
[k8s.io] Container runtime should support basic operations on container
  runtime should support listing stats for three created containers when filter is nil. [Conformance]
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:222
[BeforeEach] [k8s.io] Container
  github.com/kubernetes-sigs/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Container
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:63
[BeforeEach] runtime should support basic operations on container
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:72
[It] runtime should support listing stats for three created containers when filter is nil. [Conformance]
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:222
STEP: create first container
STEP: Get image status for image: busybox:1.28
STEP: Create container.
Jan 19 04:00:02.419: INFO: Created container "0dea6c0df0ce53a9e04a68288a199e95e22c5dda856bb1e48f796c9e0639db60"

STEP: create second container
STEP: Get image status for image: busybox:1.28
STEP: Create container.
Jan 19 04:00:02.986: INFO: Created container "dc8e532aa2846c1007e73203ebbb010b5460d418b2a4049f118d3d4a8398979a"

STEP: create third container
STEP: Get image status for image: busybox:1.28
STEP: Create container.
Jan 19 04:00:03.520: INFO: Created container "ff2c9cdc012af77dacd28de66425a31faf71e035c74470513fca38601a47014c"

STEP: start first container
STEP: Start container for containerID: 0dea6c0df0ce53a9e04a68288a199e95e22c5dda856bb1e48f796c9e0639db60
Jan 19 04:00:03.721: INFO: Started container "0dea6c0df0ce53a9e04a68288a199e95e22c5dda856bb1e48f796c9e0639db60"

STEP: start second container
STEP: Start container for containerID: dc8e532aa2846c1007e73203ebbb010b5460d418b2a4049f118d3d4a8398979a
Jan 19 04:00:03.897: INFO: Started container "dc8e532aa2846c1007e73203ebbb010b5460d418b2a4049f118d3d4a8398979a"

STEP: start third container
STEP: Start container for containerID: ff2c9cdc012af77dacd28de66425a31faf71e035c74470513fca38601a47014c
Jan 19 04:00:04.056: INFO: Started container "ff2c9cdc012af77dacd28de66425a31faf71e035c74470513fca38601a47014c"

STEP: test containers stats
STEP: List container stats for all containers:
[AfterEach] runtime should support basic operations on container
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:76
STEP: stop PodSandbox
STEP: delete PodSandbox
[AfterEach] [k8s.io] Container
  github.com/kubernetes-sigs/cri-tools/pkg/framework/framework.go:51

• [SLOW TEST:5.120 seconds]
[k8s.io] Container
github.com/kubernetes-sigs/cri-tools/pkg/framework/framework.go:72
  runtime should support basic operations on container
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:68
    runtime should support listing stats for three created containers when filter is nil. [Conformance]
    github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:222
------------------------------
SS
------------------------------
[k8s.io] Container runtime should support basic operations on container
  runtime should support listing stats for started containers [Conformance]
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:195
[BeforeEach] [k8s.io] Container
  github.com/kubernetes-sigs/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Container
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:63
[BeforeEach] runtime should support basic operations on container
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:72
[It] runtime should support listing stats for started containers [Conformance]
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:195
STEP: create container
STEP: Get image status for image: busybox:1.28
STEP: Create container.
Jan 19 04:00:06.806: INFO: Created container "2eb61d05369373530bbc4b11097b070f5398dcf717cd2e13a852dc6f717d5f36"

STEP: start container
STEP: Start container for containerID: 2eb61d05369373530bbc4b11097b070f5398dcf717cd2e13a852dc6f717d5f36
Jan 19 04:00:06.994: INFO: Started container "2eb61d05369373530bbc4b11097b070f5398dcf717cd2e13a852dc6f717d5f36"

STEP: test container stats
STEP: List container stats for all containers:
[AfterEach] runtime should support basic operations on container
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:76
STEP: stop PodSandbox
STEP: delete PodSandbox
[AfterEach] [k8s.io] Container
  github.com/kubernetes-sigs/cri-tools/pkg/framework/framework.go:51
•SSSSSSSSSS
------------------------------
[k8s.io] Container runtime should support basic operations on container
  runtime should support listing stats for started containers when filter is nil [Conformance]
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:210
[BeforeEach] [k8s.io] Container
  github.com/kubernetes-sigs/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Container
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:63
[BeforeEach] runtime should support basic operations on container
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:72
[It] runtime should support listing stats for started containers when filter is nil [Conformance]
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:210
STEP: create container
STEP: Get image status for image: busybox:1.28
STEP: Create container.
Jan 19 04:00:09.121: INFO: Created container "3ff79f6b267b676f3074b84e533900c257fa89c37dc0bc61a1842b3c6b0be8b0"

STEP: start container
STEP: Start container for containerID: 3ff79f6b267b676f3074b84e533900c257fa89c37dc0bc61a1842b3c6b0be8b0
Jan 19 04:00:09.272: INFO: Started container "3ff79f6b267b676f3074b84e533900c257fa89c37dc0bc61a1842b3c6b0be8b0"

STEP: test container stats
STEP: List container stats for all containers:
[AfterEach] runtime should support basic operations on container
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:76
STEP: stop PodSandbox
STEP: delete PodSandbox
[AfterEach] [k8s.io] Container
  github.com/kubernetes-sigs/cri-tools/pkg/framework/framework.go:51
•SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
[k8s.io] Container runtime should support basic operations on container
  runtime should support listing container stats [Conformance]
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:180
[BeforeEach] [k8s.io] Container
  github.com/kubernetes-sigs/cri-tools/pkg/framework/framework.go:50
[BeforeEach] [k8s.io] Container
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:63
[BeforeEach] runtime should support basic operations on container
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:72
[It] runtime should support listing container stats [Conformance]
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:180
STEP: create container
STEP: Get image status for image: busybox:1.28
STEP: Create container.
Jan 19 04:00:11.700: INFO: Created container "0ea95b87caaa82915540726427696219e9dde0796303806c8ca6692585897d10"

STEP: start container
STEP: Start container for containerID: 0ea95b87caaa82915540726427696219e9dde0796303806c8ca6692585897d10
Jan 19 04:00:11.892: INFO: Started container "0ea95b87caaa82915540726427696219e9dde0796303806c8ca6692585897d10"

STEP: test container stats
STEP: List container stats for containerID: 0ea95b87caaa82915540726427696219e9dde0796303806c8ca6692585897d10
[AfterEach] runtime should support basic operations on container
  github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:76
STEP: stop PodSandbox
STEP: delete PodSandbox
[AfterEach] [k8s.io] Container
  github.com/kubernetes-sigs/cri-tools/pkg/framework/framework.go:51
•SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Ran 4 of 89 Specs in 12.354 seconds
SUCCESS! -- 4 Passed | 0 Failed | 0 Pending | 85 Skipped

You're using deprecated Ginkgo functionality:
=============================================
Ginkgo 2.0 is under active development and will introduce several new features, improvements, and a small handful of breaking changes.
A release candidate for 2.0 is now available and 2.0 should GA in Fall 2021.  Please give the RC a try and send us feedback!
  - To learn more, view the migration guide at https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md
  - For instructions on using the Release Candidate visit https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#using-the-beta
  - To comment, chime in at https://github.com/onsi/ginkgo/issues/711

  You are using a custom reporter.  Support for custom reporters will likely be removed in V2.  Most users were using them to generate junit or teamcity reports and this functionality will be merged into the core reporter.  In addition, Ginkgo 2.0 will support emitting a JSON-formatted report that users can then manipulate to generate custom reports.

  If this change will be impactful to you please leave a comment on https://github.com/onsi/ginkgo/issues/711
  Learn more at: https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#removed-custom-reporters
  Measure is deprecated and will be removed in Ginkgo V2.  Please migrate to gomega/gmeasure.
  Learn more at: https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#removed-measure
    github.com/kubernetes-sigs/cri-tools/pkg/benchmark/container.go:54
    github.com/kubernetes-sigs/cri-tools/pkg/benchmark/container.go:98
    github.com/kubernetes-sigs/cri-tools/pkg/benchmark/image.go:68
    github.com/kubernetes-sigs/cri-tools/pkg/benchmark/image.go:94
    github.com/kubernetes-sigs/cri-tools/pkg/benchmark/pod.go:42
    github.com/kubernetes-sigs/cri-tools/pkg/benchmark/pod.go:86
    github.com/kubernetes-sigs/cri-tools/pkg/benchmark/pod_container.go:48

To silence deprecations that can be silenced set the following environment variable:
  ACK_GINKGO_DEPRECATIONS=1.16.5

PASS
root@ubuntu-bm:~/go/src/github.com/kubernetes-sigs/cri-tools#
```

**Comment by mikebrow** [comment]
> @minbaev the tests seems to fail, can we fix them?

hey sasha .. chatted with alexander off line.. the fix for this test case still needs the cherry pick over on containerd.. (then move to containerd 1.5.10 when it's out (wip)) or we can run this test against containerd main.. or wait for containerd 1.6.0 which is imminent and has all kinds of k/k 1.23 goodness :-)

**Review by mikebrow — APPROVED** [review]
LGTM

dependency error for the windows build is not related to this PR

we can move to containerd 1.5.10 or 1.6 when they become available, this commit off the top of 1.5 is fair to use for now

**Comment by minbaev** [comment]
@saschagrunert could you please look at the tests failing. not sure the failure is related to the changes made by this pr

https://github.com/kubernetes-sigs/cri-tools/runs/4900343760?check_suite_focus=true#step:12:112


on windows 
```
failed pulling image "[k8s.gcr.io/pause:3.1](http://k8s.gcr.io/pause:3.1)": no matching manifest for windows/amd64
```

**Comment by saschagrunert** [comment]
> @saschagrunert could you please look at the tests failing. not sure the failure is related to the changes made by this pr
> 
> https://github.com/kubernetes-sigs/cri-tools/runs/4900343760?check_suite_focus=true#step:12:112
> 
> on windows
> 
> ```
>  failed to create PodSandbox: rpc error: code = Unknown desc = failed pulling image "k8s.gcr.io/pause:3.1": no matching manifest for windows/amd64 10.0.17763 in the manifest list entries
> ```

Yes, pause is not available for windows right now. We should exclude those tests for that platform or modify them.

**Comment by saschagrunert** [comment]
@minbaev) may I ask you to rebase to get the latest CI changes in?

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Comment by saschagrunert** [comment]
The tests do not seem to work on windows. @minbaev can you take a look?
/lgtm cancel

**Comment by minbaev** [comment]
@saschagrunert Hi, what would be the mechanism to exclude the failing windows test due to "k8s.gcr.io/pause:3.1" pulling error? should it be a part of this pr?

**Comment by saschagrunert** [comment]
> @saschagrunert Hi, what would be the mechanism to exclude the failing windows test due to "k8s.gcr.io/pause:3.1" pulling error? should it be a part of this pr?

Hm, I'm wondering why it works on the master branch compared to this PR :thinking:

**Comment by mikebrow** [comment]
"critest dockershim / windows-2019 (pull_request) Failing after 7m — windows-2019" .. 

Looks to me like it's on the wrong pause tag.. 3.1 sounds .. really old probably needs 3.2 - 3.6?  dockershim tests were removed...  Gonna guess bad rebase?

**Comment by minbaev** [comment]
@mikebrow @saschagrunert thanks for looking into that
i have followed the flow to rebase with master, but still seeing that pause:3.1 issue. i'm not sure how 3.1 is set
does it maybe have to do with us setting version of containerd to 1.5.9 in the pr whereas it's 1.6.4 already

**Comment by mikebrow** [comment]
> @mikebrow @saschagrunert thanks for looking into that
> i have followed the flow to rebase with master, but still seeing that pause:3.1 issue. i'm not sure how 3.1 is set
> does it maybe have to do with us setting version of containerd to 1.5.9 in the pr whereas it's 1.6.4 already

check the list of test buckets.. there is no "critest dockershim" test in master, but you have it in your PR's test.. thus I think your PR is not based on the current master...? or I'm reading something wrong...

**Comment by minbaev** [comment]
@mikebrow there is this config to set the test
https://github.com/kubernetes-sigs/cri-tools/blob/master/.github/workflows/dockershim.yml

**Comment by minbaev** [comment]
also this pr to remove dockershim, but it isn't merged
https://github.com/kubernetes-sigs/cri-tools/pull/871

**Comment by minbaev** [comment]
should i just skip the failing tests in dockershim.yml file?

**Comment by mikebrow** [comment]
interesting.. so those are usually not triggered in PRs.. but ok they are triggered after merge..

looking at the output.. delta between you and the last merge bucket is .. your test has 11 buckets being run vs just 8 buckets.. and those 3 are failing .. see https://github.com/kubernetes-sigs/cri-tools/runs/6553793489?check_suite_focus=true#step:12:65

**Comment by mikebrow** [comment]
> should i just skip the failing tests in dockershim.yml file?

nod.. just add your new 3 buckets to the skip.. since most of them are already being skipped

or you could figure out how to move up the pause to 3.something that works.. Guessing 3.1 doesn't have the windows pause in it.. needs to be at least 3.2 would be my guess.

**Comment by mikebrow** [comment]
> also this pr to remove dockershim, but it isn't merged
> #871

you know it says dockershim but it's really Cloning into 'D:\a\cri-tools\cri-tools/src/github.com/Mirantis/cri-dockerd'

let's see... ah here it is.. and it's already being fixed https://github.com/Mirantis/cri-dockerd/pull/46

**Review by mikebrow — APPROVED** [review]
LGTM

**Review by feiskyer — APPROVED** [review]
/lgtm

**Pull request merged**
